### DISCUSSION TITLE
Small-World Sigma Definition and Source Code Inconsistency

### DISCUSSION THREAD

**Discussion by driscoll42** [networkx/networkx#5735] [Q&A]
Happy to submit a PR to correct this but before I do wanted to be sure of the exact fix that would be desired. In short, the definition for sigma in the smallworld algorithms doesn't match the code, which is not to say that it's technically incorrect.

In the sigma [documentation](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.smallworld.sigma.html) it defines sigma as (bolded my emphasis):

> The small-world coefficient is defined as: sigma = C/Cr / L/Lr where C and L are respectively the **average clustering coefficient** and average shortest path length of G. Cr and Lr are respectively the **average clustering coefficient** and average shortest path length of an equivalent random graph.

However in the [source code](https://networkx.org/documentation/stable/_modules/networkx/algorithms/smallworld.html#sigma) for sigma it actually uses transitivity:

    for i in range(nrand):
        Gr = random_reference(G, niter=niter, seed=seed)
        randMetrics["C"].append(nx.transitivity(Gr))
        randMetrics["L"].append(nx.average_shortest_path_length(Gr))

    C = nx.transitivity(G)
    L = nx.average_shortest_path_length(G)
    Cr = np.mean(randMetrics["C"])
    Lr = np.mean(randMetrics["L"])

Which is not inherently wrong as Humphries and Gurney in "[Network ‘Small-World-Ness’: A Quantitative Method for Determining Canonical Network Equivalence](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2323569/)" defined another definition of sigma using transitivity, and as the notes in the documentation specifies that they are using the Humphries et al implementation.

I think the question ultimately boils down to three options:

1. Update documentation to specify transitivity
2. Update the code to use average clustering coefficient
3. Add a new parameter to allow a user to choose which they want

I would lean toward 2, but didn't want to just make a PR with the change without asking first. The documentation could also use an example or two which would be easy enough to add as well.

And to check/confirm, [omega](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.smallworld.omega.html#networkx.algorithms.smallworld.omega) does not have this inconsistency and uses average_clustering.

**Comment by dschult** [comment]
For an undirected graph, the transitivity and average clustering should be the same, right?
And there is a very slight overhead in calling `average_clustering` because it handles all those directed cases too.

Are you getting values that are incorrect for sigma?

**Reply by driscoll42** [reply]
So, the reason I found this issue is because of a class project where we were comparing transitivity and average clustering and had to explore cases when they were different. An example could be a cycle graph where every node also connects to two inner nodes:

(An example graph)
![image](https://user-images.githubusercontent.com/4751505/173901367-74ea1c0a-9442-495f-8be2-f20a451d7ca4.png)

From going through the work, as the cycle graph nodes grows towards infinity, transitivity will approach 0 and average clustering will approaches ~ 0.833. Running this graph through a sigma using transitivity vs average clustering for n = 10 (for 12 total nodes) I get sigma of 0.99668 vs 0.994512 (transitivity vs average clustering) and then for n=100 I got sigmas of 0.99967 vs 0.99940 (transitivity vs average clustering) .

It's certainly more of an edge case, and not at all like I'm getting statistically different results, but technically they are different and mostly the inconsistency between the description and the implementation threw me off.

**Reply by driscoll42** [reply]
For some more numbers, here's the transitivity, average clustering, the averaged transitivity and clustering as defined in Sigma, Sigma values for with transitivity and average clustering using different values of n to form the cycle:

n|Transitivity|Transitivity (Avg of 10 G)|Average Clustering|Average Clustering (Avg of 10 G)|Sigma (Trans)|Sigma (Average Clustering)
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
2|1|1|1|1|1|1
5|0.75|0.75|0.785714286|0.785714286|1|1
7|0.642857143|0.648979592|0.759259259|0.77037037|0.990566038|0.985576923
10|0.529411765|0.534705882|0.755050505|0.767550505|0.99009901|0.983714427
25|0.28125|0.28125|0.78300095|0.78300095|1|1
50|0.157894737|0.157894737|0.804298643|0.804298643|1|1
100|0.08411215|0.084168224|0.817770012|0.818750404|0.999333777|0.998802575

Transitivity and average clustering are clearly different, but at least with the sample graphs I'm making the difference between the graph and the random graphs made based on it are so small as to make sigma basically the same.

**Comment by dschult** [comment]
Yes -- for sure transitivity of a graph differs from the average clustering coefficient of the graph as defined in the Strogatz Watts paper. My confusion was comparing with an older use of the term average clustering coefficient. Let's stick to the cited paper for this function as you are doing... Thanks...

In that paper they use two different measures of sigma: S^{sw} uses the average clustering coefficient, while S^{\Delta} uses transitivity.  As they state, most of their results report sigma as S^{\Delta} and so I think it makes sense that we return that value. Also, it makes sense that the differences we are seeing between the two measures is small because while transitivity and average clustering coefficient are not similar values, the ratio of their values to an average over random shifts of the graph is likely to be quite similar since they are both measures of how many triangles are in the graph.  Thus the values of sigma computed in each way are similar. 

This is clearly related to the similar question for the omega function: should it use transitivity or some sort of average clustering coefficient?  As you point out, the current code uses `nx.average_clustering_coefficient`, but this is a recent change #5216 that involved some deep dive into the literature and resulted in two changes as described [in this comment](https://github.com/networkx/networkx/issues/5064#issuecomment-975284052). Basically it switched from transitivity to avg clustering and it changed from an average to a maximum over many random perturbations of the graph of the `average_clustering_coefficient`.  That change was only made a few months ago.

@fabriz-io do you have an impression based on your reading of whether the sigma value should be computed with transitivity or average_cluster_coefficient? 

I believe that part of the difficulty here is that transitivity was (before the small-world paper) called the average clustering. That is inflated as an issue because the values of sigma do not change very much from one definition to the other.  It would be nice if we could find papers reporting on these concepts is a way that showed the importance of any such differences.
